### PR TITLE
Add IBOTPatchLoss

### PR DESCRIPTION
**Pull request by guarin**
## Changes
* Add IBOTPatchLoss
* Add Center
* Refactor DINOLoss to use Center

## How was it tested?
* Added a test that checks that values are same between our loss and the original implementation

I didn't add more tests as manually calculating test values would be very cumbersome due to the number of operations and variables involved.

### CODE REVIEW

File: lightly/models/modules/center.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        super().__init__()
+
+        modes = ["mean"]
+        self._center_fn: Callable[[Tensor], Tensor]
+        if mode not in modes:
+            raise ValueError(f"Unknown mode '{mode}'. Valid modes are {modes}.")
+        if mode == "mean":
+            self._center_fn = self._center_mean
+        else:
+            raise NotImplementedError(f"Mode '{mode}' is not implemented.")
```
Comment (MalteEbner) [inline lightly/models/modules/center.py:47]: Only the first error message will be raised. I would change this to one of these options:
```python
mode_to_center_fn = {
    "mean": self._center_mean,
}
if mode not in mode_to_center_fn:
    raise ValueError(f"Invalid mode '{mode}'. Valid modes are {list(mode_to_center_fn.keys())}.")
self._center_fn = mode_to_center_fn[mode]
```

or 
```
modes = ["mean"]
match mode:
    case "mean":
        self._center_fn = self._center_mean
    case _:
        raise ValueError(f"Invalid mode '{mode}'. Valid modes are {modes}.")
```

The first one has the advantage that you cannot accidentally forget to specify a new method at both places.

File: lightly/models/modules/center.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            (batch_size, sequence_length, feature_dim) and the center should be computed
+            across the batch and sequence dimensions, the size should be
+            (1, 1, feature_dim).
+        mode:
+            Mode to compute the center. Currently only 'mean' is supported.
+        momentum:
+            Momentum term for the center calculation.
+        _register_buffer:
+            Deprecated, do not use. This argument is only kept for backwards
+            compatibility with DINOLoss.
```
[the remark this answers is not in the document]
Comment (MalteEbner) [reply lightly/models/modules/center.py:28]: Furthermore:
- why is this True by default?
- why is the argument not ignored, but actually used?

File: lightly/models/modules/center.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        self.center.copy_(self._center_momentum(batch_center))
+
+    @torch.no_grad()
+    def _center_mean(self, x: Tensor) -> Tensor:
+        """Returns the center of the input tensor by calculating the mean."""
+        batch_center = torch.mean(x, dim=self.dim, keepdim=True)
+        if dist.is_available() and dist.is_initialized():
+            dist.all_reduce(batch_center)
+            batch_center = batch_center / dist.get_world_size()
+        return batch_center
```
Comment (MalteEbner) [inline lightly/models/modules/center.py:89]: Why is this a method if it does not use self?

File: lightly/models/modules/center.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    @torch.no_grad()
+    def _center_momentum(self, batch_center: Tensor) -> Tensor:
+        """Returns the new center with momentum update."""
+        return self.center * self.momentum + batch_center * (1 - self.momentum)
```
Comment (MalteEbner) [inline lightly/models/modules/center.py:94]: As this one-liner is only used at one place: Why not inline it?

File: tests/loss/test_ibot_loss.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        student_out = torch.tensor([[0.7, 0.8], [0.9, 1.0], [1.1, 1.2]])
+        mask = torch.tensor(
+            [[[True, False], [True, False]], [[False, False], [False, True]]]
+        )
+
+        loss = criterion.forward(
+            teacher_out=teacher_out, student_out=student_out, mask=mask
+        )
+        assert loss == pytest.approx(0.6085, rel=0.0001)
+        assert torch.all(criterion.center.value != 0)  # Check center was updated
```
Comment (MalteEbner) [inline tests/loss/test_ibot_loss.py:24]: Can't you directly assert the new center value as `(1-momentum)*torch.mean(teacher_out, dim=-1)`?

File: tests/loss/test_ibot_loss.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    def test_forward(self, device: str) -> None:
+        if not torch.cuda.is_available() and device == "cuda":
+            pytest.skip("CUDA not available")
+
+        criterion = IBOTPatchLoss(output_dim=2, teacher_temp=0.1, student_temp=0.2)
+        teacher_out = torch.tensor([[0.1, 0.2], [0.3, 0.4], [0.5, 0.6]])
+        student_out = torch.tensor([[0.7, 0.8], [0.9, 1.0], [1.1, 1.2]])
+        mask = torch.tensor(
+            [[[True, False], [True, False]], [[False, False], [False, True]]]
+        )
```
Comment (MalteEbner) [inline tests/loss/test_ibot_loss.py:18]: I am missing a test where the clamping is actually needed as there are 0 times True for one image.

**Review by MalteEbner — COMMENTED** [review]
Looks mostly good. I still have some questions about the loss design.

File: lightly/loss/ibot_loss.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    Implementation is based on [2].
+
+    - [0]: iBOT, 2021, https://arxiv.org/abs/2111.07832
+    - [1]: DINOv2, 2023, https://arxiv.org/abs/2304.07193
+    - [2]: https://github.com/facebookresearch/dinov2/blob/main/dinov2/loss/ibot_patch_loss.py
+
+    Attributes:
+        temperature:
+            Temperature of the softmax for the student outputs.
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/ibot_loss.py:20]: Ah yes, completely missed this

File: lightly/loss/ibot_loss.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        Args:
+            teacher_out:
+                Tensor with shape (B * N, D) containing the teacher output of the masked
+                tokens.
+            student_out:
+                Tensor with shape (B * N, D) containing the student output of the masked
+                tokens.
+            mask:
+                Boolean tensor with shape (B, H, W) containing the token mask.
+                Exactly B * N entries must be set to True in the mask.
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/ibot_loss.py:57]: The DINOv2 implementation doesn't use the `forward` method of the loss, it only uses the `forward_masked` method which has the following shapes:
```
        student_patch_tokens_masked: (B * N , D)
        teacher_patch_tokens_masked: (B * N, D)
        student_masks_flat: (B, H * W)
```

The only difference is the shape of the `student_masks_flat` shape. I kept it as `(B, H, W)` because this is how the mask will be shaped everywhere in the code.

File: lightly/loss/ibot_loss.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            Loss value.
+        """
+        # Calculate cross entropy loss.
+        teacher_softmax = F.softmax(
+            (teacher_out - self.center.value) / self.teacher_temp, dim=-1
+        )
+        student_log_softmax = F.log_softmax(
+            student_out / self.student_temperature, dim=-1
+        )
+        # (B * N, D) -> (B * N)
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/ibot_loss.py:69]: These comments only show the changes in the shapes of the tensors and are independent of the underlying operation.

File: lightly/loss/ibot_loss.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        loss = -torch.sum(teacher_softmax * student_log_softmax, dim=-1)
+
+        # Get weights.
+        # (B, H, W) -> (B, 1, 1)
+        num_masked_per_image = (
+            mask.sum(dim=(1, 2), keepdim=True).clamp(min=1.0).clamp(min=1.0)
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/ibot_loss.py:75]: Good catch!

File: lightly/loss/ibot_loss.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                Boolean tensor with shape (B, H, W) containing the token mask.
+                Exactly B * N entries must be set to True in the mask.
+
+        Returns:
+            Loss value.
+        """
+        # Calculate cross entropy loss.
+        teacher_softmax = F.softmax(
+            (teacher_out - self.center.value) / self.teacher_temp, dim=-1
+        )
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/ibot_loss.py:65]: It does here: https://github.com/facebookresearch/dinov2/blob/e1277af2ba9496fbadf7aec6eba56e8d882d1e35/dinov2/loss/ibot_patch_loss.py#L46

Called from here: https://github.com/facebookresearch/dinov2/blob/e1277af2ba9496fbadf7aec6eba56e8d882d1e35/dinov2/train/ssl_meta_arch.py#L200

I moved it directly into the loss calculation as it simplifies the code a lot.

File: lightly/loss/ibot_loss.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            mask.sum(dim=(1, 2), keepdim=True).clamp(min=1.0).clamp(min=1.0)
+        )
+        # (B, 1, 1) -> (B, H, W) -> (B * N)
+        weight = (1.0 / num_masked_per_image).expand_as(mask)[mask]
+
+        # Apply weighthing.
+        B = mask.shape[0]
+        loss = (loss * weight).sum() / B
+
+        self.center.update(teacher_out)
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/ibot_loss.py:84]: I would like to keep it at the end to make sure we never accidentally reuse the center after the update.

File: lightly/models/modules/center.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/models/modules/center.py:28]: > How about adding deprecation warning when the argument is used?

Then everyone using the `DINOLoss` would get a deprecation warning.

> why is this True by default?

Because you always want to register the center as a buffer. Otherwise it will not be stored in the state dict.

> why is the argument not ignored, but actually used?

Because it is used for `DINOLoss`

File: lightly/models/modules/center.py (hunk 3 above)
Comment (guarin) [reply lightly/models/modules/center.py:89]: It uses `self.dim`

File: lightly/models/modules/center.py (hunk 4 above)
Comment (guarin) [reply lightly/models/modules/center.py:94]: It leaves the option to configure it in a subclass.

File: tests/loss/test_ibot_loss.py (hunk 6 above)
Comment (guarin) [reply tests/loss/test_ibot_loss.py]: Updated, had to change the expected loss as adding more entries to the mask increases the batch size.

**Review by MalteEbner — COMMENTED** [review]
Looks almost good to merge

File: lightly/loss/ibot_loss.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+                the student output of the masked tokens.
+            mask:
+                Boolean tensor with shape (batch_size, height, width) containing the
+                token mask. Exactly batch_size * sequence_length entries must be set to
+                True in the mask.
+
+        Returns:
+            Loss value.
+        """
+        # B = batch size, N = sequence length, D = embed dim
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/ibot_loss.py]: Its actually the number of masked tokens (tokens replaced with the mask token). The terminology around masking is a bit confusing, sometimes masking means dropping tokens and sometimes replacing tokens.

File: lightly/models/modules/center.py (hunk 4 above)
Comment (guarin) [reply lightly/models/modules/center.py:93]: I would like to keep it as is.

**Review by MalteEbner — APPROVED** [review]
LGTM^2

File: lightly/models/modules/center.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/models/modules/center.py:28]: This already happens in the DINOLoss itself, it is not something the user has to configure, see: https://github.com/lightly-ai/lightly/blob/b71e1f57ff3f588c3d82d2db4d1e87efecd932c3/lightly/loss/dino_loss.py#L77
